### PR TITLE
Make getPointedThing faster

### LINKED ISSUE #3511: Setting an item's range high results in poor performance
I set an item's range to 100, and it resulted in a large increase in lag when wielded. The lag went away after I switched to something else. A range of 30 doesn't create any issues for me.

EDIT: This file defines the item in question: https://github.com/raymoo/magical_realm/blob/master/magic_spells/creative_stick.lua

### PR DESCRIPTION
**Pull request by gregorycu**
Fixes #3511

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -363,72 +442,61 @@ PointedThing getPointedThing(Client *client, Hud *hud, const v3f &player_positio
...
+	std::vector<aabb3f> boxes;
+	while (!interpolator.IsDone())
+	{
+		points.clear();
+		interpolator.Step(points);
+
+		bool hit = false;
+		for (std::vector<v3s16>::iterator iter = points.begin(); iter != points.end(); ++iter)
+		{
+			v3s16 pos = *iter;
```
Comment (Zeno-) [inline src/game.cpp]: Why?

File: src/game.cpp [hunk 2]
```diff
@@ -363,72 +442,61 @@ PointedThing getPointedThing(Client *client, Hud *hud, const v3f &player_positio
...
-	for (s16 y = ystart; y <= yend; y++) {
-		for (s16 z = zstart; z <= zend; z++) {
-			for (s16 x = xstart; x <= xend; x++) {
-				MapNode n;
-				bool is_valid_position;
-				v3s16 p(x, y, z);
+			v3s16 np = pos;
+			v3f npf = intToFloat(np, BS);
+			for (std::vector<aabb3f>::const_iterator i = boxes.begin(); i != boxes.end(); ++i) {
+				aabb3f box = *i;
```
Comment (Zeno-) [inline src/game.cpp]: Why?

Comment (gregorycu) [inline src/game.cpp]: If you're asking why I pulled it out into a variable, that's what the old code did.

Comment (gregorycu) [inline src/game.cpp]: I've remove those redundant variables.

File: src/game.cpp [hunk 3]
```diff
@@ -316,6 +316,85 @@ static inline u8 getNeighbors(v3s16 p, INodeDefManager *nodedef, ClientMap *map,
 	return neighbors;
 }
 
+namespace {
+class Interpolator
+{
+public:
+	Interpolator(v3f start, v3f end):
```
Comment (nerzhul) [inline src/game.cpp]: const v3f & for each&

File: src/game.cpp [hunk 4]
```diff
@@ -316,6 +316,85 @@ static inline u8 getNeighbors(v3s16 p, INodeDefManager *nodedef, ClientMap *map,
...
+		m_end(end),
+		m_current(start),
+		m_done(false)
+	{
+		m_step_amount = end - start;
+
+		double largest_increase_amount = fabs(m_step_amount.X);
+
+		if (fabs(m_step_amount.Y) > largest_increase_amount)
+		{
```
Comment (nerzhul) [inline src/game.cpp]: Wrong coding style

File: src/game.cpp [hunk 5]
```diff
@@ -316,6 +316,85 @@ static inline u8 getNeighbors(v3s16 p, INodeDefManager *nodedef, ClientMap *map,
...
+		if (largest_increase_amount == 0)
+		{
+			m_done = true;
+			return;
+		}
+
+		m_step_amount /= largest_increase_amount;
+	}
+
+	void Step(std::vector<v3s16>& points)
```
Comment (est31) [inline src/game.cpp]: Never ever do non-const references. They are forbidden by the code style and are generally a very bad idea.

Comment (est31) [inline src/game.cpp]: (At least when passing them through functions, as normal variables they are ofc okay).

File: src/game.cpp [hunk 6]
```diff
@@ -316,6 +316,85 @@ static inline u8 getNeighbors(v3s16 p, INodeDefManager *nodedef, ClientMap *map,
 	return neighbors;
 }
 
+namespace {
```
Comment (est31) [inline src/game.cpp]: Why is this required?

File: src/game.cpp [hunk 7]
```diff
@@ -363,72 +442,58 @@ PointedThing getPointedThing(Client *client, Hud *hud, const v3f &player_positio
...
-	if (xend == 32767)
-		xend = 32766;
+	v3f start = camera_position / BS;
+	v3f end = start + (camera_direction * d);
+
+	Interpolator interpolator(start, end);
+
+	std::vector<v3s16> nodes;
+	std::vector<aabb3f> boxes;
+	while (!interpolator.IsDone())
```
Comment (est31) [inline src/game.cpp]: The code style rule is: braces always on the same line, except for function braces. I've seen this at multiple places in the PR, but only make one comment about it.

File: src/game.cpp [hunk 8]
```diff
@@ -316,6 +316,85 @@ static inline u8 getNeighbors(v3s16 p, INodeDefManager *nodedef, ClientMap *map,
 	return neighbors;
 }
 
+namespace {
+class Interpolator
```
Comment (est31) [inline src/game.cpp]: Not a single line of documentation about this class and its methods? Thats bad, especially as the name is so generic.

**Comment by est31** [comment]
I'm not sure, does this account for nodes with selection boxes that exceed their nodes? We should at least support selection boxes that leave their originating node by two nodes.

**Comment by gregorycu** [comment]
@est31 Why two nodes? The existing code only works with one node. Yes, it does account for selection boxes that are one larger than the current node.

File: src/game.cpp (hunk 6 above)
Comment (gregorycu) [inline src/game.cpp]: Otherwise it will have external linkage and have the potential to clash with classes with the same name.

File: src/game.cpp (hunk 5 above)
Comment (gregorycu) [inline src/game.cpp]: I don't agree with it. But don't care enough to argue.

File: src/game.cpp (hunk 7 above)
Comment (gregorycu) [inline src/game.cpp]: Ok. Will fix.

File: src/game.cpp (hunk 8 above)
Comment (gregorycu) [inline src/game.cpp]: The name is generic, but the scope is limited. The class exists to pull out the Interpolation state from the function itself (my first iteration had this all getPointedThing but it seemed to be a lot for the reader to hold in their head at once.) Given the name, and the usage, it seems to me to be pretty clear what it does.

**Comment by gregorycu** [comment]
I have updated the commit.

File: src/game.cpp [hunk 9]
```diff
@@ -363,72 +438,57 @@ PointedThing getPointedThing(Client *client, Hud *hud, const v3f &player_positio
...
-					pointed_pos = np;
-				}
+				result.type = POINTEDTHING_NODE;
+				min_distance = distance;
+				pointed_pos = pos;
+				hit = true;
 			}
-		}
+		}
+		if (hit)
```
Comment (kwolekr) [inline src/game.cpp]: Can't you put this in the while condition instead?

File: src/game.cpp (hunk 8 above)
Comment (est31) [inline src/game.cpp]: Still a comment would help people alot :)

File: src/game.cpp (hunk 6 above)
Comment (est31) [inline src/game.cpp]: Ah, okay

**Comment by Fixer-007** [comment]
Didn't used extralong range with it, just standard creative and noncreative and found no problems with it.

**Comment by SmallJoker** [comment]
~~With pull~~ Before/Raw:
screenshot_20160506_194627_
~~Before/Raw~~ With pull:
screenshot_20160506_194813_

**Comment by gregorycu** [comment]
It's not too obvious in the screenie, but in the second pic the distance is much further away.

**Comment by gregorycu** [comment]
I have updated the code to reflect all feedback (style, moving checks, adding comments). @est31 @kwolekr @Zeno-

**Comment by ShadowNinja** [comment]
I don't like how this still uses the old "move forward a bit and round" method.  You could take a look at the algorithm that the mining laser uses, that one works pretty well is natively node-based.

**Comment by kwolekr** [comment]
Will be looking at this.  Sorry for the delay.

**Comment by gregorycu** [comment]
This is not a naive "move forward a bit and round". It moves "forward" by the smallest amount required to get a different node. Are you able to link the mining laser mod so I can see what it does?

**Comment by ShadowNinja** [comment]
@gregorycu: This is the best algorithm I've managed to get so far: https://github.com/minetest-technic/technic/blob/2b6abfe733a7d44ac4a444e91f2b4b47e51b553d/technic/helpers.lua#L66

It tracks two positions: one is a floating point position which is precisely moved down the ray slowly; the other one is the integer position, which is adjusted by moving it one node in whichever direction is farthest from the current floating position on the ray.  This is repeated until the end is reached.

The simple step-and-round method can still miss things if the ray only hits the edge of a box and the step distance isn't small enough to catch it.

**Comment by gregorycu** [comment]
Thanks SN. I appreciate the comments.

The existing code calculates that step size (it's actually a step vector where one of the dimensions is 1 and the others are all less than one). Each step is guaranteed to have at least one node. The step size has no effect on the correctness of the result set. Only the number of false positives returned.

When more than one dimension changes (in the edge scenario you mentioned) all diagonal nodes are returned as well. So they will not be skipped.

Theoretically i could make the step size the range of the tool and I would be just as correct, except I would return a lot more nodes. 

I'll have to study the algo you posted. Just wanted to let you know that mine won't miss nodes, your algo may still be better.

**Comment by raymoo** [comment]
Is this still active?

**Comment by SmallJoker** [comment]
The current state of this pull works fine, I'm using these changes in my Win32 builds for over a month now and it works fine. I really hope this pull will either get merged soon or slightly improved to make it even faster.

File: src/game.cpp [hunk 10]
```diff
@@ -363,72 +447,56 @@ PointedThing getPointedThing(Client *client, Hud *hud, const v3f &player_positio
...
-						continue;
-					}
-					result.type = POINTEDTHING_NODE;
-					min_distance = distance;
-					pointed_pos = np;
-				}
+				result.type = POINTEDTHING_NODE;
+				min_distance = distance;
+				pointed_pos = pos;
+				hit = true;
```
Comment (SmallJoker) [inline src/game.cpp]: Why was here the `break;` left away? That would prevent from checking all left nodes in the vector.

**Comment by est31** [comment]
#4346 was merged, which makes this PR obsolete.
